ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/manager/EthServerTest.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.manager;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.singletonList;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import tech.pegasys.pantheon.ethereum.chain.Blockchain;
+import tech.pegasys.pantheon.ethereum.core.Hash;
+import tech.pegasys.pantheon.ethereum.db.WorldStateArchive;
+import tech.pegasys.pantheon.ethereum.eth.messages.GetNodeDataMessage;
+import tech.pegasys.pantheon.ethereum.eth.messages.NodeDataMessage;
+import tech.pegasys.pantheon.util.bytes.BytesValue;
+
+import java.util.Optional;
+
+import org.junit.Before;
+import org.junit.Test;
+
+public class EthServerTest {
+
+  private static final BytesValue VALUE1 = BytesValue.of(1);
+  private static final BytesValue VALUE2 = BytesValue.of(2);
+  private static final BytesValue VALUE3 = BytesValue.of(3);
+  private static final Hash HASH1 = Hash.hash(VALUE1);
+  private static final Hash HASH2 = Hash.hash(VALUE2);
+  private static final Hash HASH3 = Hash.hash(VALUE3);
+  private final Blockchain blockchain = mock(Blockchain.class);
+  private final WorldStateArchive worldStateArchive = mock(WorldStateArchive.class);
+  private final EthPeer ethPeer = mock(EthPeer.class);
+  private final EthMessages ethMessages = new EthMessages();
+
+  @Before
+  public void setUp() {
+    new EthServer(blockchain, worldStateArchive, ethMessages, 2);
+  }
+
+  @Test
+  public void shouldRespondToNodeDataRequests() throws Exception {
```

Comment:
The other tests of this flavor are in `EthProtocolManagerTest` - we should probably consolidate them in one place.  I originally put these tests on `EthProtocolManager` so that the test were less closely tied to implementation details (the serving logic being segregated in this internal package-private `EthServer` class).  I have a slight preference for keeping them there, but I'll leave it up to you.

Reply:
The `EthProtocolManagerTest` cases are kind of more integration tests - they cover quite a large chunk of system. Good to have as they ensure it all fits together well but for testing the detail like that limits are applied and missing data is skipped I generally find it's better to focus on a smaller unit which EthServer works well for.

I've added a test in `EthProtocolManagerTest` for `getNodeData` to ensure it does all hang together properly.   

I suspect if I had to change the other request types much in the future I'd likely write tests in `EthServerTest` for them just because they're smaller and easier to work with. `EthProtocolManagerTest` is ensuring they work well now though so I don't feel particularly compelled to go add them now. I'm happy to if you feel strongly about it though.

Reply:
I do tend to prefer higher-level tests for this stuff, but I am not super opinionated.  So, lgtm - merge away!